args/core/compute/kernel.hpp

```diff
@@ -0,0 +1,183 @@
+             * @note this needs to be done for all buffers before the kernel is dispatched
+             * @param buffer The buffer you want to inform the kernel about
+             * @param name The name of the Buffer you want to inform the kernel about
+             */
+            Kernel& set_buffer(Buffer buffer, const std::string& name);
+
+            /**
+             * @brief Sets a Buffer as the Kernel Argument
+             * @note this is required for all Kernel Arguments regardless of buffer direction
+             * @note this needs to be done for all buffers before the kernel is dispatched
+             * @param buffer The buffer you want to inform the kernel about
+             * @param index The index of the Buffer you want to inform the kernel about
+             */
+            Kernel& set_buffer(Buffer buffer, cl_uint index);
+
+
+            /**
+             * @brief same as informing the kernel about a buffer and then enqueueing it 
+             */
+            Kernel& set_and_enqueue_buffer(Buffer buffer, block_mode blocking = block_mode::BLOCKING);
+
+            /**
+             * @brief same as informing the kernel about a buffer and then enqueueing it 
+             */
+            Kernel& set_and_enqueue_buffer(Buffer buffer, const std::string&, block_mode blocking = block_mode::BLOCKING);
+
+            /**
+             * @brief same as informing the kernel about a buffer and then enqueueing it 
+             */
+            Kernel& set_and_enqueue_buffer(Buffer buffer, cl_uint index, block_mode blocking = block_mode::BLOCKING);
+
+            /**
+             * @brief Dispatches the Kernel to the command-queue
+             * @note All Kernel Parameters must be set before this can be called
+             * @note global and local must be set before this can be called
+             * @pre set_buffer
+             * @pre set_and_enqueue_buffer
+             * @pre local
+             * @pre global
+             */
+            Kernel& dispatch();
+
+            /**
+             * @brief Ensures that the command-queue is committed and finished executing
+             * @pre dispatch
+             */
+            void finish() const;
+
+        private:
+
+            buffer_type m_default_mode;
+            std::map<std::string, cl_uint> m_paramsMap;
+            Program* m_prog;
+            cl_kernel m_func;
+            cl_command_queue m_queue;
+            size_t m_global_size;
+            size_t m_local_size;
+
+
+            //helper function to wrap parameter checking 
+            template <class F,class... Args>
+            void param_find(F && func,std::string name)
+            {
+                // lazy build buffer names
+                if (m_paramsMap.empty())
+                {
+                    build_buffer_names();
+                }
+
+                //find and do thing, otherwise print warning
+                if (const auto it = m_paramsMap.find(name); it != m_paramsMap.end())
```

Comment:
if you're using a map anyways, why use std::find? can't you use the accessor operator or .at()?